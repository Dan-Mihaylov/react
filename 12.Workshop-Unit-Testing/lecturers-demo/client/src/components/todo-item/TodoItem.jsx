import React from 'react'

const TodoItem = ({
    _id,
    text,
    isCompleted,
    onToggle,
}) => {
    return (
        <li className="todo-item py-4">
            <div className="flex items-center">
                <input
                    id="todo1"
                    name="todo1"
                    type="checkbox"
                    checked={isCompleted}
                    onChange={() => onToggle(_id)}
                    className="h-4 w-4 text-teal-600 focus:ring-teal-500 border-gray-300 rounded"
                />
                <label htmlFor="todo1" className="ml-3 block text-gray-900">
                    <span className="text-lg font-medium">{text}</span>
                </label>
            </div>
        </li>
    )
}

export default TodoItem
