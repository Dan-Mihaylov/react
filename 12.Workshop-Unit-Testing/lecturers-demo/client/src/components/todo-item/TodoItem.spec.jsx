import { it, expect, beforeEach, vi } from 'vitest'
import ReactDOM from 'react-dom/client'
import { act } from 'react'
import { render, screen, fireEvent } from '@testing-library/react'

import TodoItem from './TodoItem'

// Custom way
beforeEach(() => {
    document.body.innerHTML = '';
});

// Custom way
it('Should display todo text', async () => {
    const container = document.createElement('div');

    document.body.appendChild(container);

    const root = ReactDOM.createRoot(container);

    await act(() => {
        root.render(<TodoItem text="t1" />)
    });

    const textElement = document.querySelector('label[for=todo1] span');

    expect(textElement.textContent).toEqual('t1');
});

// Testing lirbary way
it('Should display todo text', () => {
    render(<TodoItem text="test1" />);

    const textElement = screen.queryByText('test1');

    expect(textElement).toBeInTheDocument();
});

it('Should be checked when isCompleted', () => {
    render(<TodoItem isCompleted text="text1" />);

    const checkbox = screen.getByRole('checkbox', { name: 'text1' })

    expect(checkbox).toBeChecked();
});

it('Should be checked when clicked', () => {
    const onToggle = vi.fn();

    render(
        <TodoItem
            _id="123"
            isCompleted={false}
            text="checkbox-1"
            onToggle={onToggle}
        />
    );

    const checkbox = screen.getByRole('checkbox', { name: 'checkbox-1' });

    fireEvent.click(checkbox);

    expect(onToggle).toBeCalled();
    expect(onToggle).toBeCalledTimes(1);
    expect(onToggle.mock.calls[0]).toEqual(['123'])
});

it('Should match snapshot', () => {
    render(
        <TodoItem
            _id="123"
            isCompleted={false}
            text="checkbox-1"
        />
    );

    expect(document.body).toMatchSnapshot();
});
