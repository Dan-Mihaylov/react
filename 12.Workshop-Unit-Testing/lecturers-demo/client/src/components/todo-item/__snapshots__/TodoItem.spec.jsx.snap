// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should match snapshot 1`] = `
<body>
  <div>
    <li
      class="todo-item py-4"
    >
      <div
        class="flex items-center"
      >
        <input
          class="h-4 w-4 text-teal-600 focus:ring-teal-500 border-gray-300 rounded"
          id="todo1"
          name="todo1"
          type="checkbox"
        />
        <label
          class="ml-3 block text-gray-900"
          for="todo1"
        >
          <span
            class="text-lg font-medium"
          >
            checkbox-1
          </span>
        </label>
      </div>
    </li>
  </div>
</body>
`;
